### PR TITLE
Time-series aggregation

### PR DESCRIPTION
**Pull request by ivanmalagon**
Related to: https://github.com/CartoDB/cartodb/issues/12324

Multi PR. Changes in 
cartodb PR: https://github.com/CartoDB/cartodb/pull/12347
cartodb.js PR: https://github.com/CartoDB/cartodb.js/pull/1685
deep-insights.js PR: https://github.com/CartoDB/deep-insights.js/pull/555

**BRAND NEW REFURBISHED TIME-SERIES WIDGETS!!** 📊 

The time-series now aggregate data by date if it's based on a date column.

image

timeseries

----

⚠️  **IMPORTANT NOTES ABOUT ACCEPTANCE PROCESS!!** ⚠️ 

You **must** deploy Windshaft PR https://github.com/CartoDB/Windshaft-cartodb/pull/698 in order to get Maps API updated.

Before deploying this PR **PLEASE**:
1. Select any target environment but ded07. We need a clean server and we've already deployed there to check some things.
1. Go to that target environment and create several maps with combinations of time-series, animated time-series and histograms. We need to check that the changes don't break existing widgets.
1. Then you're ready to start deploying.

----

**WHAT TO ASSERT**

We have to check that the widget behaves. Using a dataset with date columns (please find below one that you can use) check:

CREATION

- Create a time-series widget from everywhere it can be created:
  - From widget tab > time-series > a date column. (A)
  - From widget tab > time-series > a numeric column. (B)
  - From widget tab > histogram > a numeric column. (C)
  - Within a layer > data tab > select a date column. (D)
  - Within a layer > change style to animated. (E)

A. Date time-series widget from the beginning.
  - Check that aggregation selection is empty. This means that the aggregation is automatic.
  - Once the widget is loaded, change aggregation and check that graphic changes accordingly.
  - Change column in the form, to numeric, to another date and back.
  - When changing to another date column, check that the available possible aggregations change, according to the data domain of that column.

B. Numeric time-series
  - Change bins. Widget should reflect that number of bins.
  - Change column.

C. Numeric histogram.
  - In widgets tab, change type to time-series.
  - Check that widget is well created.

D. From data tab
  - Check that the created widget is aggregated by date.

E. Animated
  - Change selected column from cartodb_id to a date column. It should show time aggregated data.

OLD WIDGETS
Go to the maps that you should have been created with a myriad of widgets and check that they are well instantiated.

FILTER
Having a date time-series widget.
- Select several bins.
- Check that axis labels are meaningful, related to the aggregation.
- Selected bins should reflect the correspondant data in the map.

CHANGES IN SQL
There is a corner case (perhaps not that corner) that can ruin the map.
- Apply an analysis (or set a SQL query) to limit the extent of a column date data. For instance, if your column have data for a whole year, apply a SQL query to limit it to one day.
- Create a time-series widget. Since your data only extends to one day, you are given very granular aggregations, like day, hour and minute. Select hour, for instance.
- The widget now is set to aggregate by hour.
- Revert the SQL to get again a whole year.
- If we aggregrate now by minute, we'd have more than 500.000 buckets.
- The widget should show an error and not break the map.

EMBEDS
Please check that everything works as well in embed views. For every widget you create, test it in embeds. Test in embeds also the old widgets that have not been modified.

MISC
Please, act like a chaos monkey and try to break the widgets by any means.

----

**DATASETS**
Any dataset with date columns will suffice. Anyway, this is a dataset that has two date columns: `posted_time` and `actor_postedtime`so you can try switching date columns.

[Dataset with two date columns](https://team.carto.com/u/alex-carto/dataset/twitter_chiquito_calzada_chiquigif)

### CODE REVIEW

**Comment by ivanmalagon** [comment]
We've deployed this PR to ded07 to check several things. Still pending Code Review and some testing but you can start playing with time aggregations in staging, @noguerol @saleiva @kevin-reilly , to check that everything fits the product vision.

The selection of timezone is scheduled for another ticket.

As I said, it's in ded07. If you need further credentials ping me in private.

**Comment by ivanmalagon** [comment]
Found
image
2014 out of bounds. I know that we look to the future but this is too much 😉

**Comment by rubenmoya** [comment]
I've found another problem:

**Steps to reproduce**
- Select a date column
- Filter the widget
- Change the column to a numeric one

**Expected**
The filter is reset since the column is different.

**Current**
The filters keeps active with the previous bin indexes.

**Comment by kevin-reilly** [comment]
I'm trying to test this feature using the jvillar-ded04 user and the times series widget is not rendering.  Here is my map:

https://jvillar-ded04.carto-staging.com/builder/b895c5c1-71af-46db-8254-e93a04de2786

screen shot 2017-07-17 at 12 50 57 pm

**Comment by matallo** [comment]
spotted here, too https://github.com/CartoDB/cartodb/pull/12482#issuecomment-316684144

File: lib/assets/core/javascripts/cartodb3/data/widget-definitions-collection.js [hunk 1]
```diff
@@ -21,46 +29,44 @@ module.exports = Backbone.Collection.extend({
...
 
   initialize: function (models, opts) {
-    if (!opts.configModel) throw new Error('configModel is required');
-    if (!opts.mapId) throw new Error('mapId is required');
-    if (!opts.layerDefinitionsCollection) throw new Error('layerDefinitionsCollection is required');
-
-    this._configModel = opts.configModel;
-    this._mapId = opts.mapId;
-    this._layerDefinitionsCollection = opts.layerDefinitionsCollection;
+    var self = this;
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/data/widget-definitions-collection.js]: why not using `.bind(this)` ?

File: webpack.dev.config.js [hunk 2]
```diff
@@ -33,7 +33,7 @@ module.exports = env => {
         resolve(__dirname, 'lib/assets/node_modules')
       ]
     },
-    devtool: 'cheap-module-eval-source-map',
+    devtool: 'eval-source-map',
```
Comment (matallo) [inline webpack.dev.config.js]: until now the I got the best results with `source-map` but I don't think that's important

**Comment by kevin-reilly** [comment]
I spotted an issue that occurs when attempting to create multiple widgets in separate steps, not at the same time.

I am using the `sandyrelated311` dataset.  

### Steps to reproduce

1. Create a new map
2. Click add widget
3. Add time series widget using created_date field
4. Click continue
5. Click add widget to add your second widget
6. Add category widget using agency_name

The second widget does not appear.  It will appear if you refresh the browser.

Add both widgets at the same time appears to work correctly.

**Comment by kevin-reilly** [comment]
Now I'm embarrassed to admit how much time I spent reproducing it to
document it correctly.

:-)

--
*Kevin Reilly*
SVP Product
*CARTO <https://carto.com/> | kevin@carto.com <kevin@carto.com>*
(917) 375-2168

**Comment by rjimenezda** [comment]
Great job @ivanmalagon @rubenmoya & everyone else involved, what a cool feature 🔝 💯 💥 

Some stuff I've found so far:

Selecting a range on a numeric time series and changing the number of buckets breaks with `Uncaught TypeError: Cannot read property 'start' of undefined` on time-series-widget-model:37.

On romanded01 there's a map (storms) with a dataset that allows aggregations by minutes, but when you select it, the widget gets this 'refresh' overlay and it just doesn't work. I see no failing requests. This date comes from a to_timestamp call.

I'll keep going, just wanted to leave these two for now

**Comment by ivanmalagon** [comment]
I'll start working in those two issues found by @rjimenezda

**Comment by rjimenezda** [comment]
Couldn't find anything else, but I would encourage more people to give it a try 👌

**Comment by matallo** [comment]
just realized mobile histograms go out of the width bounds

pageshot of nurburgring laps import 2017-07-04-1905 05

**Comment by ivanmalagon** [comment]
Very good catch @rjimenezda 

The query we were using for guessing the most appropiate aggregations had one syntax error that have only shown up because you were using the `to_timestamp` function.

Apart from that, we were using the total difference in days to do the calculations but your dataset has a difference of 0 days, allowing all kind of aggregations. But the dataset had a difference of 1409 of minutes and the widget fails gracefully.

Now I'm trying to set another query that solve the problem.

One more thing, did you test the widgets also in embed pages and with previous existing time-series?

**Comment by rjimenezda** [comment]
Happy little coincidence, I am only using to_timestamp because I don't have that many datasets with dates in them 😄 

Yes, I checked the embeds, although I did not check all responsive versions to be honest.

I created most of the maps/widgets before deploying, in order to test nothing was broken, and then created a few more using all the different possibilities, switching types, etc.

**Comment by ivanmalagon** [comment]
Fixed 3 issues. The two reported by Roman and one that happened after setting an aggregation for a column, then switching to another column but letting the aggregation in blank and then refreshing. The widget definition still had the first one, the one set by the user. Normally, it is not a big deal but if the user selected an aggregation not available in the second column (minutes, for instance), then the widget would appear broken after refreshing.

Now I'm going to resolve the conflicts with the latest deep-insights-integrations changes before going for another round of acceptance.

**Comment by rjimenezda** [comment]
New code looks good, and the issues I reported are gone 👋

**Comment by ivanmalagon** [comment]
retest this please

**Comment by ivanmalagon** [comment]
We've made a change on how to instantiate the map and how we make calls to the dataview URL.

Now, the parameter `aggregation` is mandatory when dealing with `date` columns. In case you want the backend to choose the aggregation for you, you must set the param to `auto`.

Why? Because not sending anything and depending solely on the column type could potentially break maps made by partners. If they didn't send the `bins` params, the API would choose the proper bins number. With the latest changes, in case that was made to a `date` column, the API would return another payload, suitable for the new date aggregations.

So now, the API sends numeric payload if no param is requested. If you want date aggregations, the `aggregation` param is mandatory.

**Comment by ivanmalagon** [comment]
One problem spotted.

In the document [Time-series milestone](https://docs.google.com/document/d/1rPgCk8BiHXsNM4BQPb82ImQzVzs0UTTiWXF8vsuCXRw/edit#heading=h.eilx4ykmr119) the list of possible aggregations by the API goes from `minute` to `year`. We've spotted that in real life, the API is sending also `second` but the frontend is not prepared right now for that. We don't go that far for aggregations (based on what was written in the document).

I'm making the minimum change for that. At least the axis tips show the time text but all related to the widget form starts from `minute`.

Pinging @dgaubert How far does the backend go? Second? Millisecond?

**Comment by xavijam** [comment]
Hey guys, be careful with submodules changes. Both were pointing to old commits.

**Comment by xavijam** [comment]
Deploying to ded05 again, just in case.

**Comment by xavijam** [comment]
Working properly :dancer:, great job guys :muscle:.

**Pull request merged**